slatedb/src/db.rs

```diff
@@ -588,6 +595,50 @@ impl Db {
         Ok(())
     }
 
+    /// Create a snapshot of the database.
+    ///
+    /// ## Returns
+    /// - `Result<Arc<DbSnapshot>, Error>`: the snapshot of the database, it represents
+    ///   a consistent view of the database at the time of the snapshot.
+    ///
+    /// ## Examples
+    ///
+    /// ```
+    /// use slatedb::{Db, Error};
+    /// use slatedb::object_store::{ObjectStore, memory::InMemory};
+    /// use std::sync::Arc;
+    /// use bytes::Bytes;
+    ///
+    /// #[tokio::main]
+    /// async fn main() -> Result<(), Error> {
+    ///     let object_store = Arc::new(InMemory::new());
+    ///     let db = Db::open("test_db", object_store).await?;
+    ///     
+    ///     // Write some data and create a snapshot
+    ///     db.put(b"key1", b"value1").await?;
+    ///     let snapshot = db.snapshot().await?;
+    ///     
+    ///     // Snapshot provides read-only access to database state
+    ///     let value = snapshot.get(b"key1").await?;
+    ///     assert_eq!(value, Some(Bytes::from(b"value1".as_ref())));
+    ///     
+    ///     // Write more data to original database
+    ///     db.put(b"key2", b"value2").await?;
+    ///     
+    ///     // Snapshot still sees old state, original db sees new data
+    ///     assert_eq!(snapshot.get(b"key2").await?, None);
+    ///     assert_eq!(db.get(b"key2").await?, Some(Bytes::from(b"value2".as_ref())));
+    ///     
+    ///     Ok(())
+    /// }
+    /// ```
+    pub async fn snapshot(&self) -> Result<Arc<DbSnapshot>, crate::Error> {
+        self.inner.check_error()?;
+        let seq = self.inner.oracle.last_seq.load();
```

Comment:
doesn't this need to be the last committed seq num? Otherwise, the snapshot may observe the before and after of in-flight writes (both a current on-going write and the next write).

e.g. the following sequence:
(1) the write task is in the middle of writing a write with seq 10, but hasn't applied writes. In this case last_seq is now 11
(2) a caller R creates a db snapshot S (with seq_num 11)
(3) R reads from S
(4) the write task commits the writes with seq_num 10
(5) R reads from S
(6) the write task commits the next write (with seq_num 11)
(7) R reads from S

Those 3 reads will observe different states

Reply:
oh yes, this is a serious bug, thank you for this. let me fix it and consider about how to cover a test case for it.

Reply:
updated with a test case `test_snapshot_with_committed_failpoint` with failpoint to reproduce the commit seq issue, PTAL